### ISSUE TITLE
Future of ISerializable, Serializable attribute and IDeserializationCallback interface

### ISSUE THREAD

**Issue by georgiuk** [dotnet/runtime#70208]
After reading the [Binary Formatter Obsoletion Strategy](https://github.com/dotnet/designs/blob/main/accepted/2020/better-obsoletion/binaryformatter-obsoletion.md) , I have some questions.
It seems that the ISerializable interface will become obsolete in .NET 7.0 and will be removed completely in .NET 8.0. Correct? 
However from what I understand the [Serializable] attribute will stay. Correct?
Also, there is no mention on the future of IDeserializationCallback interface. What will happen to it?

Thank you

**Comment by danmoseley** [comment]
@GrabYourPitchforks

**Comment by georgiuk** [comment]
So, will all implementations of ISerializable contained within .NET throw exception?
Also what about the IDeserializationCallback implementations?

**Comment by huoyaoyuan** [comment]
I don't think so. The "infrastructure" means `BinaryFormatter` and its contents.

**Comment by georgiuk** [comment]
Please check the highlighted sentences. What I understand is that what will be marked as obsolete in .NET 7.0 will throw exception in 8.0 regardless if it is BinaryFormatter or "BinaryFormatter infrastructure". And I think ISerializable could be considered BinaryFormatter infrastructure.  
![image](https://user-images.githubusercontent.com/1830368/173146930-b29e9df5-d6d7-4a67-b8aa-00c379deae3f.png)

**Comment by GrabYourPitchforks** [comment]
We're running a little behind on our obsoletion strategy because we've had other higher-priority work items come up in the meantime. I plan on updating the document soon to reflect more achievable timelines.

The `[Serializable]` and `[NonSerialized]` attributes are overall part of the `ISerializable` infrastructure which is being obsoleted. However, there are several non-BinaryFormatter related serialization technologies which consume those attributes independent of other parts of the `ISerializable` infrastructure. So we're still investigating the overall impact on the ecosystem before we make a determination as to the ultimate fate of those types. Nonetheless, these are considered legacy attributes, and we don't expect future serialization technologies to honor these attributes.

**Comment by jeffhandley** [comment]
@GrabYourPitchforks has updated the [BinaryFormatter obsoletion document](https://github.com/dotnet/designs/blob/main/accepted/2020/better-obsoletion/binaryformatter-obsoletion.md#binaryformatter-infrastructure-removed-from-net-net-9) to reflect the revised timeline.

The portion relevant to this issue is included in the [.NET 9 timeframe](https://github.com/dotnet/designs/blob/main/accepted/2020/better-obsoletion/binaryformatter-obsoletion.md#binaryformatter-infrastructure-removed-from-net-net-9):

> In .NET 9.0, the entirety of the BinaryFormatter infrastructure will be removed from the product. Here, "BinaryFormatter infrastructure" refers to any API which was obsoleted (as error) as part of an earlier stage of this proposal, even if that API is not itself on the BinaryFormatter type. The implementation of all such APIs will be to throw PlatformNotSupportedException. The internal BinaryFormatter implementation will no longer exist in .NET in any form and cannot be reenabled through any compat switch. Functionality may be moved to an unsupported out-of-band package.

By changing the implementations to throw and exception, this will address the type loader concern above.

@georgiuk, does this address your concerns at this point? Please feel free to share more information about your specific scenarios. It could help us to understand more about how you're using the BinaryFormatter-related APIs.

**Comment by georgiuk** [comment]
Thank you for the answer. 
It is clearer now. 
Some missing points:
What happens with the IDeserializationCallback interface?
Will the marked-as-warning APIs throw Exception on .NET 9.0?
